### PR TITLE
Issue #12164: Enable Checker Framework

### LINKED ISSUE #12164: Enable Checker Framework
In our relentless pursuit to use all static analysis and code quality tools, we would like to add the [Checker Framework](https://checkerframework.org/) to our project. This framework uses annotations and utilities to enhance type checking and reduce potential defects in code.

Here is the built-in Checkers:

 - [ ] [Nullness Checker](https://checkerframework.org/manual/#nullness-checker) for null pointer errors 
- [ ] [Initialization Checker](https://checkerframework.org/manual/#initialization-checker) to ensure all @NonNull fields are set in the constructor 
- [ ] [Map Key Checker](https://checkerframework.org/manual/#map-key-checker) to track which values are keys in a map 
- [ ] [Optional Checker](https://checkerframework.org/manual/#optional-checker) for errors in using the [Optional](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html) type 
- [ ] [Interning Checker](https://checkerframework.org/manual/#interning-checker) for errors in equality testing and interning 
- [ ] [Called Methods Checker](https://checkerframework.org/manual/#called-methods-checker) for the builder pattern 
- [ ] [Resource Leak Checker](https://checkerframework.org/manual/#resource-leak-checker) for ensuring that resources are disposed of properly 
- [ ] [Fake Enum Checker](https://checkerframework.org/manual/#fenum-checker) to allow type-safe fake enum patterns and type aliases or typedefs 
- [ ] [Tainting Checker](https://checkerframework.org/manual/#tainting-checker) for trust and security errors 
- [ ] [Lock Checker](https://checkerframework.org/manual/#lock-checker) for concurrency and lock errors 
- [ ] [Index Checker](https://checkerframework.org/manual/#index-checker) for array accesses 
- [ ] [Regex Checker](https://checkerframework.org/manual/#regex-checker) to prevent use of syntactically invalid regular expressions 
- [ ] [Format String Checker](https://checkerframework.org/manual/#formatter-checker) to ensure that format strings have the right number and type of % directives 
- [ ] [Internationalization Format String Checker](https://checkerframework.org/manual/#i18n-formatter-checker) to ensure that i18n format strings have the right number and type of {} directives 
- [ ] [Property File Checker](https://checkerframework.org/manual/#propkey-checker) to ensure that valid keys are used for property files and resource bundles 
- [ ] [Internationalization Checker](https://checkerframework.org/manual/#i18n-checker) to ensure that code is properly internationalized 
- [ ] [Signature String Checker](https://checkerframework.org/manual/#signature-checker) to ensure that the string representation of a type is properly used, for example in Class.forName 
- [ ] [GUI Effect Checker](https://checkerframework.org/manual/#guieffect-checker) to ensure that non-GUI threads do not access the UI, which would crash the application 
- [ ] [Units Checker](https://checkerframework.org/manual/#units-checker) to ensure operations are performed on correct units of measurement 
- [ ] [Signedness Checker](https://checkerframework.org/manual/#signedness-checker) to ensure unsigned and signed values are not mixed 
- [ ] [Purity Checker](https://checkerframework.org/manual/#purity-checker) to identify whether methods have side effects 
- [ ] [Constant Value Checker](https://checkerframework.org/manual/#constant-value-checker) to determine whether an expression’s value can be known at compile time 
- [ ] [Reflection Checker](https://checkerframework.org/manual/#reflection-resolution) to determine whether an expression’s value (of type Method or Class) can be known at compile time 
- [ ] [Initialized Fields Checker](https://checkerframework.org/manual/#initialized-fields-checker) to ensure all fields are set in the constructor 
- [ ] [Aliasing Checker](https://checkerframework.org/manual/#aliasing-checker) to identify whether expressions have aliases 
- [ ] [Must Call Checker](https://checkerframework.org/manual/#must-call-checker) to over-approximate the methods that should be called on an object before it is de-allocated 
- [ ] [Subtyping Checker](https://checkerframework.org/manual/#subtyping-checker) for customized checking without writing any code 
- [ ] [Third-party checkers](https://checkerframework.org/manual/#third-party-checkers) that are distributed separately from the Checker Framework 

Not all of these apply to our project, we can discuss the Checkers here and decide which will be valuable for us. Also, suppression of various Checkers can be a bit different, so we can take this on a case-by-case basis.

1 Checker == 1 PR

As for minimizing/ keeping annotations out of our codebase, we can consider https://checkerframework.org/annotation-file-utilities/#viewing-source, which is a tool to externally store annotations and insert them into source code as needed (such as during execution of Checker).

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #12164

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -2382,6 +2389,382 @@
...
+      <id>checker-framework-index</id>
+      <build>
+        <plugins>
+          <plugin>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <version>${maven.compiler.plugin.version}</version>
+            <configuration>
+              <!-- Must fork or else JVM arguments are ignored. -->
+              <fork>true</fork>
+              <meminitial>1024m</meminitial>
```
Comment (nrmancuso) [inline pom.xml:2543]: Please add meminital and maxmem properties to all profiles.

Comment (Vyom-Yadav) [reply pom.xml:2576]: Done.

**Comment by Vyom-Yadav** [comment]
Error from https://github.com/checkstyle/checkstyle/actions/runs/3047499328/jobs/4911550442

`
Error:  /home/runner/work/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java:[40,16] error: [type.checking.not.run] OptionalChecker did not run because of a previous error issued by javac
`

Running checker framework without `-Awarns` issues a lots of warnings like these.
From [SO Answer](https://stackoverflow.com/a/70139615/14425335) (By creator of Checker Framework) 

> The Checker Framework runs as a plugin to javac. When javac issues an error in one class (including any Checker Framework error), javac may or may not process other classes. Unfortunately, there is no good way for a user to predict how far javac will get.

Creating a different checker from each profile will also not work. If a checker fails then new warnings from that checker can also be like `did not run because of a previous error issued by javac`.

Using `-Awarns` will simply convert errors to warnings.

**Comment by romani** [comment]
Usage of warns might help only to not let maven execution fail with error code. So it will help us to analyze results and do failure based on some other reasons.

**Comment by Vyom-Yadav** [comment]
> Usage of warns might help only to not let maven execution fail with error code.

That is not the main purpose of using `-Awarns`, not failing on error can be done by
 `<failOnError>false</failOnError>`.

We are using `-Awarns` because many errors are there due to the checker not running due to 
`a previous error issued by javac.` Using `-Awarns` only shows the error we need, without the errors of type `did not run because of a previous error issued by javac.`

**Comment by romani** [comment]
Good to know.

Can you briefly answer my other question ?

**Comment by Vyom-Yadav** [comment]
> Do we have problems to get stable output from Checker execution ? I try to understand what blocks it us to merge this PR

I executed it locally a couple of times, and new violations pop up in the CI, trying once more.

File: .ci/StarterRuleSet-AllRulesByCategory.groovy.txt [hunk 2]
```diff
@@ -359,9 +359,6 @@ ruleset {
     NonFinalSubclassOfSensitiveInterface
     ObjectFinalize
     PublicFinalizeMethod
-    SystemExit {
-        doNotApplyToFileNames = 'pitest-survival-check-xml.groovy,error-prone-check.groovy'
-    }
```
Comment (Vyom-Yadav) [inline .ci/StarterRuleSet-AllRulesByCategory.groovy.txt:364]: We suppress every groovy script, removal of this is better, we will most probably use groovy only for writing scripts.

**Comment by romani** [comment]
@Vyom-Yadav , please put in commit proper issue number - https://github.com/checkstyle/checkstyle/issues/12164 

please fix codenarc

File: .ci/checker-framework-suppressions/checker-framework-nullness-optional-interning-suppressions.xml [hunk 3]
```diff
@@ -0,0 +1,6090 @@
...
+  <checkerFrameworkError unstable="true">
+    <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/HandlerFactory.java</fileName>
+    <specifier>argument</specifier>
+    <message>incompatible argument for parameter constructor of invokeConstructor.</message>
+    <lineContent>handlerCtor, indentCheck, ast, parent);</lineContent>
+    <details>
+      found   : @Initialized @Nullable Constructor&lt;capture#788 extends @Initialized @Nullable Object&gt;
+      required: @Initialized @NonNull Constructor&lt;capture#788 extends @Initialized @Nullable Object&gt;
+    </details>
+  </checkerFrameworkError>
```
Comment (Vyom-Yadav) [inline .ci/checker-framework-suppressions/checker-framework-nullness-optional-interning-suppressions.xml:2973]: Unstable violation

File: .ci/checker-framework-suppressions/checker-framework-nullness-optional-interning-suppressions.xml [hunk 4]
```diff
@@ -0,0 +1,6090 @@
...
+  <checkerFrameworkError unstable="true">
+    <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/HandlerFactory.java</fileName>
+    <specifier>type.argument</specifier>
+    <message>incompatible type argument for type parameter T extends Object of invokeConstructor.</message>
+    <lineContent>resultHandler = (AbstractExpressionHandler) CommonUtil.invokeConstructor(</lineContent>
+    <details>
+      found   : capture#788[ extends @UnknownKeyFor Object super @KeyForBottom Void]
+      required: [extends @UnknownKeyFor Object super @UnknownKeyFor NullType]
+    </details>
+  </checkerFrameworkError>
```
Comment (Vyom-Yadav) [inline .ci/checker-framework-suppressions/checker-framework-nullness-optional-interning-suppressions.xml:3414]: Unstable violation.

**Review by romani — APPROVED** [review]
Ok let's try this Checker in read day to day coding

**Review by nrmancuso — APPROVED** [review]
There are a lot of nuances to understand with each checker, let’s merge this and live with it in order to learn. @Vyom-Yadav please create an issue to resolve checker suppressions, and in it please explain details of unstable suppressions.

**Pull request merged**